### PR TITLE
`Development`: Update eslint

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation and Context
Update to the latest version

### Description
Only code review is needed

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
- [ ] Code Review 3
- [ ] Code Review 4

## Summary by CodeRabbit

- **New Features**
  - Introduced comprehensive ESLint setup tailored for TypeScript, Angular, Prettier, Jest, and Jest-Extended.
  - Added drag-enter listener for mobile devices in the drag-and-drop question component.

- **Bug Fixes**
  - Improved drag-and-drop functionality on mobile devices.

- **Refactor**
  - Reorganized dependencies in the navbar component.
  - Updated variable declarations to `const` in `index.html` for better readability.
  - Refined conditional logic in various components and API requests.

- **Chores**
  - Updated package versions and added new dependencies in `package.json`.
  - Adjusted linting scripts for enhanced code quality checks.
  - Removed unnecessary `eslint-disable-next-line` comments.

### CODE REVIEW

File: src/main/webapp/app/exercises/programming/shared/instructions-render/programming-exercise-instruction.component.ts [hunk 1]
```diff
@@ -233,7 +233,7 @@ export class ProgrammingExerciseInstructionComponent implements OnChanges, OnDes
     updateMarkdown(): void {
         // make sure that always the correct result is set, before updating markdown
         // looks weird, but in setter of latestResult are setters of sub components invoked
-        // eslint-disable-next-line no-self-assign
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/shared/instructions-render/programming-exercise-instruction.component.ts:236]:
```suggestion
```

File: src/main/webapp/app/exercises/quiz/manage/quiz-exercise-update.component.ts [hunk 2]
```diff
@@ -556,7 +556,7 @@ export class QuizExerciseUpdateComponent extends QuizExerciseValidationDirective
             return [];
         }
         // Release Date valid but lies in the past
-        // eslint-disable-next-line no-constant-condition
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/quiz/manage/quiz-exercise-update.component.ts:559]:
```suggestion
```

File: src/main/webapp/app/exercises/quiz/shared/questions/drag-and-drop-question/drag-and-drop-question.component.ts [hunk 3]
```diff
@@ -19,10 +19,11 @@ polyfill({
 });
 
 // Drag-enter listener for mobile devices: without this code, mobile drag and drop will not work correctly!
+// eslint-disable-next-line @typescript-eslint/no-unused-expressions
 (event: any) => {
     event.preventDefault();
 };
-/* eslint-enable */
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/quiz/shared/questions/drag-and-drop-question/drag-and-drop-question.component.ts:26]:
```suggestion
```

File: src/main/webapp/app/overview/course-tutorial-groups/course-tutorial-group-card/course-tutorial-group-card.component.ts [hunk 4]
```diff
@@ -7,7 +7,7 @@ import { Course, isMessagingEnabled } from 'app/entities/course.model';
     selector: 'jhi-course-tutorial-group-card',
     templateUrl: './course-tutorial-group-card.component.html',
     styleUrls: ['./course-tutorial-group-card.component.scss'],
-    // eslint-disable-next-line @angular-eslint/no-host-metadata-property
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/overview/course-tutorial-groups/course-tutorial-group-card/course-tutorial-group-card.component.ts:10]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 5]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
                     // reference closing tag: [/referenceType] (wrapped between 3 characters)
                     // referenceStr: string to be displayed for the reference
                     // linkToReference: link to be navigated to on reference click
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:116]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 6]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
...
                     referenceStr = this.content.substring(this.content.indexOf(']', patternMatch.startIndex)! + 1, this.content.indexOf('(', patternMatch.startIndex)!);
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:118]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 7]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
...
                     linkToReference = [this.content.substring(this.content.indexOf('(', patternMatch.startIndex)! + 1, this.content.indexOf(')', patternMatch.startIndex))];
                 } else if (ReferenceType.ATTACHMENT === referenceType || ReferenceType.ATTACHMENT_UNITS === referenceType) {
                     // referenceStr: string to be displayed for the reference
                     // attachmentToReference: location of attachment to be opened on reference click
                     // attachmentRefDir: directory of the attachment
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:124]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 8]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
...
                     referenceStr = this.content.substring(this.content.indexOf(']', patternMatch.startIndex)! + 1, this.content.indexOf('(', patternMatch.startIndex)!);
                     const attachmentRefDir = this.ATTACHMENT_DIR;
                     attachmentToReference =
-                        // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
                         attachmentRefDir + this.content.substring(this.content.indexOf('(', patternMatch.startIndex)! + 1, this.content.indexOf(')', patternMatch.startIndex));
                 } else if (ReferenceType.SLIDE === referenceType) {
                     // referenceStr: string to be displayed for the reference
                     // slideToReference: location of attachment to be opened on reference click
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:132]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 9]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
...
                     const attachmentUnitRefDir = this.ATTACHMENT_DIR;
                     slideToReference =
-                        attachmentUnitRefDir +
-                        // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
-                        this.content.substring(this.content.indexOf('(', patternMatch.startIndex)! + 1, this.content.indexOf(')', patternMatch.startIndex));
+                        attachmentUnitRefDir + this.content.substring(this.content.indexOf('(', patternMatch.startIndex)! + 1, this.content.indexOf(')', patternMatch.startIndex));
                 } else if (ReferenceType.USER === referenceType) {
                     // referenceStr: string to be displayed for the reference
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:139]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 10]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
...
                     referenceStr = this.content.substring(this.content.indexOf(']', patternMatch.startIndex)! + 1, this.content.indexOf('(', patternMatch.startIndex)!);
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:141]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 11]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
...
                     queryParams = {
                         referenceUserLogin: this.content.substring(this.content.indexOf('(', patternMatch.startIndex)! + 1, this.content.indexOf(')', patternMatch.startIndex)),
                     } as Params;
                 } else if (ReferenceType.CHANNEL === referenceType) {
                     // referenceStr: string to be displayed for the reference
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:147]:
```suggestion
```

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 12]
```diff
@@ -113,43 +113,40 @@ export class PostingContentComponent implements OnInit, OnChanges, OnDestroy {
...
                     referenceStr = this.content.substring(this.content.indexOf(']', patternMatch.startIndex)! + 1, this.content.indexOf('(', patternMatch.startIndex)!);
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:149]:
```suggestion
```

File: src/test/cypress/support/requests/ExerciseAPIRequests.ts [hunk 13]
```diff
@@ -318,7 +318,7 @@ export class ExerciseAPIRequests {
             assessmentDueDate: dayjsToString(assessmentDueDate),
         };
         let newModelingExercise;
-        // eslint-disable-next-line no-prototype-builtins
+
```
Comment (Strohgelaender) [inline src/test/cypress/support/requests/ExerciseAPIRequests.ts:321]:
```suggestion
```

File: src/test/cypress/support/requests/ExerciseAPIRequests.ts [hunk 14]
```diff
@@ -415,7 +415,7 @@ export class ExerciseAPIRequests {
         const dates = {
             releaseDate: dayjsToString(releaseDate),
         };
-        // eslint-disable-next-line no-prototype-builtins
+
```
Comment (Strohgelaender) [inline src/test/cypress/support/requests/ExerciseAPIRequests.ts:418]:
```suggestion
```

File: src/test/playwright/support/requests/ExerciseAPIRequests.ts [hunk 15]
```diff
@@ -322,7 +322,7 @@ export class ExerciseAPIRequests {
             assessmentDueDate: dayjsToString(assessmentDueDate),
         };
         let newModelingExercise;
-        // eslint-disable-next-line no-prototype-builtins
+
```
Comment (Strohgelaender) [inline src/test/playwright/support/requests/ExerciseAPIRequests.ts:325]:
```suggestion
```

**Review by Strohgelaender — COMMENTED** [review]
There are some places where the removed eslint-disable comments created empty lines, I marked them all below:

**Review by pzdr7 — APPROVED** [review]
Code looks good 👍 . Maybe consider removing the newlines @Strohgelaender mentioned

File: src/main/webapp/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.ts [hunk 16]
```diff
@@ -166,6 +167,7 @@ export class AdminImportStandardizedCompetenciesComponent {
                 this.importCount.knowledgeAreas -= 1;
                 this.dataSource.data = this.importData.knowledgeAreas.map((knowledgeArea) => convertToKnowledgeAreaForTree(knowledgeArea));
             }
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (JohannesStoehr) [inline src/main/webapp/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.ts:170]: We can adapt the no-unused-vars rule, so errors can be unused but other variables still have to:
https://eslint.org/docs/latest/rules/no-unused-vars#caughterrors

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.ts (hunk 16 above)
Comment (krusche) [reply src/main/webapp/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.ts:170]: good idea, can we define this in the global eslint config?

Comment (krusche) [reply src/main/webapp/app/admin/standardized-competencies/import/admin-import-standardized-competencies.component.ts]: found it and applied it

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by Strohgelaender — APPROVED** [review]

File: .dockerignore [hunk 17]
```diff
@@ -18,7 +18,6 @@ www/
 # do not exclude .git as it's necessary to insert the commit id into the build
 !.git
 # do not exclude .eslint files as they are required in the build process
-!.eslintignore
 !.eslintrc
```
Comment (Jan-Thurner) [inline .dockerignore:21]: Can this line get removed too, as the File got deleted in this PR?
```suggestion
```

**Pull request merged**